qutebrowser/misc/sql.py

```diff
@@ -0,0 +1,178 @@
+    return query
+
+
+class SqlTable(QObject):
+
+    """Interface to a sql table.
+
+    Attributes:
+        Entry: The class wrapping row data from this table.
+        _name: Name of the SQL table this wraps.
+        _primary_key: The primary key of the table.
+
+    Signals:
+        changed: Emitted when the table is modified.
+    """
+
+    changed = pyqtSignal()
+
+    def __init__(self, name, fields, primary_key, parent=None):
+        """Create a new table in the sql database.
+
+        Raises SqlException if the table already exists.
+
+        Args:
+            name: Name of the table.
+            fields: A list of field names.
+            primary_key: Name of the field to serve as the primary key.
+        """
+        super().__init__(parent)
+        self._name = name
+        self._primary_key = primary_key
+        run_query("CREATE TABLE {} ({}, PRIMARY KEY ({}))".format(
+            name, ','.join(fields), primary_key))
+        # pylint: disable=invalid-name
+        self.Entry = collections.namedtuple(name + '_Entry', fields)
+
+    def __iter__(self):
+        """Iterate rows in the table."""
+        result = run_query("SELECT * FROM {}".format(self._name))
+        while result.next():
+            rec = result.record()
+            yield self.Entry(*[rec.value(i) for i in range(rec.count())])
+
+    def __contains__(self, key):
+        """Return whether the table contains the matching item.
+
+        Args:
+            key: Primary key value to search for.
+        """
+        query = run_query("SELECT * FROM {} where {} = ?".format(
+            self._name, self._primary_key), [key])
+        return query.next()
+
+    def __len__(self):
+        """Return the count of rows in the table."""
+        result = run_query("SELECT count(*) FROM {}".format(self._name))
+        result.next()
+        return result.value(0)
+
+    def __getitem__(self, key):
+        """Retrieve the row matching the given key.
+
+        Args:
+            key: Primary key value to fetch.
+        """
+        result = run_query("SELECT * FROM {} where {} = ?".format(
+            self._name, self._primary_key), [key])
+        result.next()
+        rec = result.record()
+        return self.Entry(*[rec.value(i) for i in range(rec.count())])
+
+    def delete(self, value, field=None):
+        """Remove all rows for which `field` equals `value`.
+
+        Args:
+            value: Key value to delete.
+            field: Field to use as the key, defaults to the primary key.
+
+        Return:
+            The number of rows deleted.
+        """
+        field = field or self._primary_key
+        query = run_query("DELETE FROM {} where {} = ?".format(
+            self._name, field), [value])
+        if not query.numRowsAffected():
+            raise KeyError('No row with {} = "{}"'.format(field, value))
+        self.changed.emit()
+
+    def insert(self, values, replace=False):
```

Comment:
yes! I meant to do that after 572aaf1b4. It could be `self.insert(self.Entry(field, field, ...), replace=True)`